### PR TITLE
need to impl swap

### PR DESCRIPTION
**Pull request by jacoobes**
- swap replaces a previous dependency with a new one
     - calls `dispose` hook on the old dependency
     - removes the hook from the hook list
     - adds respective init & dispose hooks

### CODE REVIEW

File: packages/ioc/src/container.ts [hunk 1]
```diff
@@ -82,8 +82,18 @@ export class Container {
 
         const existing = this.__singletons.get(key);
         if (!existing) {
+            // @jacoobes should we return false if no existing key?
```
Comment (jacoobes) [inline packages/ioc/src/container.ts:85]: yes

File: packages/ioc/src/container.ts [hunk 2]
```diff
@@ -82,8 +82,18 @@ export class Container {
...
             throw Error("No existing key to swap for " + key);
         }
+        // check if there's dispose hook, and call it
+        if (hasCallableMethod(existing, 'dispose')) {
+            this.disposeAll().then(() => {;
```
Comment (jacoobes) [inline packages/ioc/src/container.ts:90]: this.disposeAlll would call dispose on all objects, not intended

Comment (jacoobes) [reply packages/ioc/src/container.ts:90]: This was what I was stuck on, how would we call dispose on existing object, delete it from hooks array, and then add existing?

Comment (EvolutionX-10) [reply packages/ioc/src/container.ts:90]: will see

**Pull request merged**